Reduce usage of RPC providers via relying on them sparingly

tukantje opened the pull request:
# Current Situation
We are currently relying on a `JsonRpcProvider`, in every instance, via Infura.

# What
This PR includes a couple of proposals. Code quality can definitely be improved, this iteration is mainly for discussion of the ideas.

* Use RPC providers only when we cannot setup a provider based off of `window.ethereum`.
  * This would allow us to not hit infura, for said users. Might be beneficial.
  * Due to reduced calls, our usage of infura would go down significantly.
  * Certain calls are "free" when done in this fashion - such as `eth_chainId` (Based on `ethers.js` discussion here https://github.com/ethers-io/ethers.js/issues/901)
* Use `InfuraProvider` instead of `JsonRpcProvider` for infura endpoints.
  * `ethers.js` claims this would include optimisations on their end, specifically for infura.
* Use `StaticJsonRpcProvider` instead of `JsonRpcProvider` for non-infura cases.
  * The difference is essentially making the `eth_chainId` call.
  * This is useful for when the `chainId` could change on the fly. However, the way we use these endpoints, it does seem like that is not a possibility (i.e. we create a provider for Gnosis Chain, specifically).

# Why
Main point for this was trying to figure out amount of calls we are making to infura, as it has been a recent source of pain. During this process, I've noticed that `eth_chainId` and `eth_blockNumber` are the highest hit calls, main reason being this information needing to be polled. Specifically,
* Every 10-15 secs, we poll for `eth_blockNumber`.
  * This is in line with what Infura suggests.
* Every call we make, we verify `eth_chainId`.
  * `StaticJsonRpcProvider` would stop making this call, albeit definitely should be discussed.
  * Relying on wallets instead of RPC would make this a cost-free call.
  * Main reason this is being called all the time is due to EIP-155.

Which made me also think about if we could reduce our reliance in general. A recent Slack discussion (https://cowservices.slack.com/archives/C036G0J90BU/p1679062770503109) was related to this topic, as such I've started exploring this PR.

# Important Points
I've mainly tested this out as;
* One browser with Metamask
* One browser without any wallet

After which I have mainly tested;
* Can I load the application?
* Can I connect using Wallet Connect?
* Can I send a transaction?

Which seems to be working, however this is nowhere near enough. I presume, if we desire such a change, we should;
* Test all wallets
* Think about a list of functionality that must pass instead of what I could think about randomly

Thank you, and looking forward to hearing what everybody thinks 🐮

alfetopito commented:
The test did not go so well for me
Latest FF with a bunch of tracker blockers on - which may or may not be the root cause - the behaviour is the same as when the infura key was blocked. The page does not fully load

![image](https://user-images.githubusercontent.com/43217/230345841-8fb931e4-121f-44c3-8f30-d7c6f87dc939.png)

On the console, there are various failures:
![image](https://user-images.githubusercontent.com/43217/230345771-1755e2df-00a6-4614-a398-f0726909d0cc.png)
![image](https://user-images.githubusercontent.com/43217/230346031-cb90c80f-cb68-49fd-82c9-1f462b987529.png)

On Chrome though, where I have no blockers enabled, the page loads

As you mentioned, changing the network in the app while not connected is not possible, the app breaks

While connected using WalletConnect, it works fine, balances load.

Have not tried to submit any tx to test, though.

---

In summary, if the connection without a wallet can be fixed, we should ship it

tukantje commented:
@alfetopito thank you for the review!

So multiple things are at play here from what I can see;

> Application doesn't open
Most likely this is due to not being on the same default chain (i.e. your wallet was on Goerli / Gnosis and not Mainnet).
> WalletConnect does work
WC apparently expects RPC URLs to be given to it.

I've updated the PR - apologies for the size of it.

What I've addressed here are;

* Try to centralise where RPC configuration is coming from
  * I think this part is not done yet, however it should be an improvement. Not having a centralised setup makes things quite confusing, and tough to debug (i.e. "why am I sending a call to infura when I set this to alchemy? oh wait in this file I also have an infura link")

* Return a list of RPC providers instead of singular ones, to have a fallback style handling
  * Initially I've started with creating a custom provider to handle all these cases. However, after quite a bit of research and pain, I've noticed that suggested approach on this is to handle it at a higher level than `ethers` (i.e. `web3-react`). As such, I've tried to rely on their functionality on this as much as possible.
  * `web3-react` does allow you to give multiple providers, and use them in order.

* Add public providers as a last resort

* Add sentry logging

* Fix signing issues regarding API based `JsonRpcProviders`
  * Providers for Infura, Alchemy etc. are apparently made using `UrlJsonRpcProvider` which results in not having a signer present.

* Remove the assumption of user arriving with mainnet selected on their wallet
  * This was not an issue earlier due to `chainId` checks galore.

What do you think?

tukantje commented on src/custom/constants/networks.ts:
```diff
@@ -28,3 +38,142 @@ export const RPC_URLS: { [key in SupportedChainId]: string } = {
+ *
+ * Why not instantiate and return a JsonRpcSigner? Because it has a constructor guard that prevents it from being instantiated directly.
+ * Intention with this guard is to prevent users from accidentally using a JsonRpcSigner with another provider. Which is not possible here.
+ */
```
We've shortly chatted about this part - thought I'd add context here as well.

Current behaviour is that we create a `JsonRpcProvider`, which has the following `getSigner` function ([source](https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/providers/src.ts/json-rpc-provider.ts#L488-L490)).

```
getSigner(addressOrIndex?: string | number): JsonRpcSigner {
    return new JsonRpcSigner(_constructorGuard, this, addressOrIndex);
}
```

This works quite well for us so far. However, all these API based providers are instantiated from abstract class `UrlJsonRpcProvider` which has the following implementation ([source](https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/providers/src.ts/url-json-rpc-provider.ts#L81-L87));

```
getSigner(address?: string): JsonRpcSigner {
        return logger.throwError(
            "API provider does not support signing",
            Logger.errors.UNSUPPORTED_OPERATION,
            { operation: "getSigner" }
        );
    }
```

That is problematic. And it is wrong - verifiably so, because we are able to sign messages after all and all we use is an `InfuraProvider` on our current codebase.

What if we created our own `Signer` then? Well, that also turns out to be problematic because, the available options are;

* Create a `Wallet` instance
  * This is suggested by pretty much all API providers
  * It cannot work for us! `Wallet` instances need your private key
* Create a `VoidSigner`
  * These are readonly signers that cannot sign anything
  * We need to be able to sign things
* Create a `JsonRpcSigner`
  * Not possible, due to the `constructorGuard` above.
  * This is at the heart of why we override through prototype - we don't have a way of working around the `constructorGuard` otherwise (short of patching the package)
  * In discussions it is stated that the intention with this guard is to avoid "create a signer for a provider, and then try to use it with another provider" scenario; which would be a massive footgun. It also does not apply to us here so we should be safe.
* Create a new `Signer` from scratch
  * This would be quite a piece of code, and it would pretty much replicate `JsonRpcSigner`
  * Maintaining / reviewing it I think is not worth avoiding this override

I hope this gives enough context about this part.

shoom3301 replied:
>We need to be able to sign things

Yes, but only when a wallet is connected. In this situation, we use signer as:
```
const { provider } = useWeb3React()
  const signer = provider.getSigner()
```

Maybe I'm wrong, but it looks like a wallet provider is responsible for signing and we don't need it for JsonRpcSigner, do we?

shoom3301 replied:
Have you tried to sign an order with using `VoidSigner`?

tukantje replied:
I haven't since it was defined as ([source](https://docs.ethers.org/v5/api/signer/));
> A VoidSigner is a simple Signer which cannot sign.

> Yes, but only when a wallet is connected. In this situation, we use signer as:
>Maybe I'm wrong, but it looks like a wallet provider is responsible for signing and we don't need it for JsonRpcSigner, do we?
Underneath, `web3-react` gets the provider from its connectors. [These providers store providers we supply here internally](https://github.com/Uniswap/web3-react/blob/main/packages/network/src/index.ts#LL74C5-L77C8), as such when you call for `provider.getSigner()` you are actually calling the `getSigner` method of one of the providers created. (Alternatively, if you supply only string links, it creates `JsonRpcProvider` in place. However, that is not super desirable in all cases, e.g. public providers, optimisations etc.)

In most cases? That will be fine. Proposed changes try to rely on your wallet first anyways.

However once that is not the case, you end up hitting `getSigner` method of `UrlJsonRpcProvider` which throws an error.

shoom3301 replied:
>However once that is not the case, you end up hitting getSigner method of UrlJsonRpcProvider which throws an error.

Is it possible in case when any wallet is connected?

tukantje replied:
In this iteration, that'd be dependent on whether or not `window.ethereum` is present or not. To my understanding, in WalletConnect cases that wouldn't be there, creating that possibility.

alfetopito commented on src/cow-react/modules/wallet/web3-react/connection/network.tsx:
```diff
@@ -1,12 +1,13 @@
 import { Network } from '@web3-react/network'
 
-import { RPC_URLS } from 'constants/networks'
+// import { RPC_URLS } from 'constants/networks'
```
Remove rather than comment out

tukantje replied:
Good point, clearing

alfetopito commented on src/custom/constants/networks.ts:
```diff
@@ -28,3 +38,142 @@ export const RPC_URLS: { [key in SupportedChainId]: string } = {
+export const RPC_CONFIG: { [key in SupportedChainId]: { chainId: SupportedChainId; name: string } } = {
+  [SupportedChainId.MAINNET]: {
+    chainId: SupportedChainId.MAINNET,
+    name: 'homestead',
```
Hum?

alfetopito commented on src/custom/constants/networks.ts:
```diff
@@ -28,3 +38,142 @@ export const RPC_URLS: { [key in SupportedChainId]: string } = {
+
+  result.pop()
+
+  if (RPC_TYPE[chainId] === 'infura') {
+    result.push(new InfuraProvider(RPC_CONFIG[chainId], INFURA_KEY))
+  } else {
+    result.push(new StaticJsonRpcProvider(RPC_URLS[chainId]))
+  }
```
`RPC_URLS` contains infura urls for mainnet & goerli, and the default rpc for gnosis.

Does it make sense to use it here?
Intuitively, it would only ever be used for gnosis chain.

Might as well no longer export it and rename it to something more meaningful, like fallback rpc urls, static rpc urls, etc.

What do you think?

tukantje replied:
It is a good point, although this was essentially a shot on my side for generalisation.

Simplified it - I think now we should be in a better shape.

alfetopito commented on src/cow-react/modules/wallet/web3-react/connection/coinbase.tsx:
```diff
@@ -38,7 +38,7 @@ const [web3CoinbaseWallet, web3CoinbaseWalletHooks] = initializeConnector<AsyncC
               actions,
               options: {
-                url: RPC_URLS[SupportedChainId.MAINNET],
+                url: getRpcUrls(SupportedChainId.MAINNET)[0],
```
Why only the first item? Before it was getting the full array if I read this right

tukantje replied:
Previously this was getting a single URL because RPC_URLS was only returning a single URL.

Coinbase expects a single URL unfortunately, and doesn't support an array of it as far as I can tell.

anxolin replied:
ubnernit: just add a small comment saying sth  like

`// Set first RPC url from mainnet (coinbase is limited to use only one RPC)`

alfetopito reviewed (approved):
A couple of questions in the code, but overall it looks quite good.
The explanations help understand what's going on.

As for the behaviour, briefly tested on FF with multiple networks while disconnected, connected with metamask and via wallet connect.
It all worked great 👍

shoom3301 commented on src/cow-react/modules/wallet/web3-react/connection/network.tsx:
```diff
@@ -1,12 +1,13 @@
 
 const [web3Network, web3NetworkHooks] = initializeConnector<Network>(
-  (actions) => new Network({ actions, urlMap: RPC_URLS, defaultChainId: 1 })
+  (actions) => new Network({ actions, urlMap: PROVIDERS })
```
Why did you delete defaultChainId? I don't have any concerns, just curious

shoom3301 commented on src/custom/constants/networks.ts:
```diff
@@ -28,3 +38,142 @@ export const RPC_URLS: { [key in SupportedChainId]: string } = {
+    result.push(new Web3Provider(window.ethereum, RPC_CONFIG[chainId]))
+  }
+
+  result.pop()
```
Why do we do `pop`? Looks like it just cancel 105 line

tukantje replied:
Oh no! My fault - it was a bit too late I guess. It was a debug line to see everything else working as expected. Removed it.

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+}
+
+export const PROVIDERS_RPC_URLS_FIRST_ONLY: Record<SupportedChainId, string> = {
+  [SupportedChainId.MAINNET]: PROVIDERS_RPC_URLS[SupportedChainId.MAINNET][0],
```
nit: this `PROVIDERS` and `PROVIDERS_RPC_URLS` is a bit repetitive. Considere using an IIFE returning the 3 consts, the function would just `reduce` the list of supported chains

anxolin replied:
I don't think we need this, right? we already have `PROVIDERS`?

anxolin commented on src/cow-react/modules/wallet/web3-react/connection/ledger.tsx:
```diff
@@ -26,7 +26,7 @@ const [ledger, ledgerHooks] = initializeConnector<AsyncConnector>(
                 actions,
                 options: {
-                  rpc: RPC_URLS,
+                  rpc: PROVIDERS_RPC_URLS_FIRST_ONLY,
```
nit: we could add the comment in the same line as the one of coinbase explaining that ledger allows just one RPC per network

anxolin commented on src/cow-react/modules/wallet/web3-react/connection/network.tsx:
```diff
@@ -1,12 +1,12 @@
-import { RPC_URLS } from 'constants/networks'
 import { ConnectionType } from '@cow/modules/wallet'
 import { Web3ReactConnection } from '../types'
+import { PROVIDERS } from '@src/custom/constants/networks'
```
drop the `@src/custom` part

tukantje replied:
Done

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+  PocketProvider,
+  AnkrProvider,
+} from '@ethersproject/providers'
+import * as Sentry from '@sentry/react'
```
Fror this `networks` file, i think we should move it to `cow-react` maybe wait until my refactor is done. When we "refactor/centralise" sth, is good we remove it from `custom` and put it in the right place of `cow-react`

tukantje replied:
It is a good point - I'll update in place for now and wait for your refactor to be finished to move them around.

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
- * These are the network URLs used by the interface when there is not another available source of chain data
+ * We use this class for type inference.
+ *
+ * UrlJsonRpcProvider is an abstract class, therefore we have trouble inferring the type of its constructor.
```
`we have trouble` or `can't`?

i would think if it's abstract, we don't know the constructor of the children. Is this what you are refering to?

tukantje replied:
I wrote it with my downton abbey glasses apparently 😅 Will rephrase, you are correct, we can't.

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+  [SupportedChainId.GNOSIS_CHAIN]: getProvider(SupportedChainId.GNOSIS_CHAIN),
+}
+
+export const getRpcUrls = (chainId: SupportedChainId) => PROVIDERS[chainId].map((provider) => provider.connection.url)
```
Why exporting this, plus the `consts`? I would think one thing or the other, right?

tukantje replied:
This exports URLs specifically instead of the provider itself. So for example, when you are using the public provider, we can give you the URL. It is there to support wallets that expect a URL instead of a provider instance, while still allowing for a public provider to be injected.

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+  },
+}
+
+function constructSentryError(baseError: unknown, { message }: { message: string }) {
```
nit: potencially can be abstracted into something more generic

anxolin replied:
maybe some JS doc would be good here.

tukantje replied:
Added some comments. I think we can abstract it, but I've stolen it from another part in code; so would love to handle that as a follow up to not mix up refactoring those parts with this change.

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+    errorType: 'provider',
+  }
+
+  return { baseError, sentryError: constructedError, tags }
```
why returning also the baseError that is not modified and part of the params?

tukantje replied:
Removed it, good point.

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+
+      Sentry.captureException(sentryError, { tags })
+
+      throw new Error(message)
```
Why do you create a sentry error, but throw another error? `sentryError` also has the message, wouldn't be simpler to just throw that?

tukantje replied:
Updated this to use the same instance, good point.

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+
+export const providers = {
+  get mainnet() {
+    const _providers = getProvider(SupportedChainId.MAINNET)
```
you have already a const where you pre-calculated this, why not using it? `PROVIDERS[SupportedChainId.MAINNET]`?

tukantje replied:
Updated it, good point.

anxolin commented on src/cow-react/types.ts:
```diff
@@ -8,6 +8,16 @@ export type Writeable<T> = { -readonly [P in keyof T]: T[P] }
+*/
+export type Newable<T extends new (...args: any) => any> = {
+  new (...args: ConstructorParameters<T>): InstanceType<T>
+}
```
cool

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+  [SupportedChainId.GNOSIS_CHAIN]: PROVIDERS_RPC_URLS[SupportedChainId.GNOSIS_CHAIN][0],
+}
+
+export const providers = {
```
why this providers are lowercase?

anxolin commented on src/custom/constants/networks.ts:
```diff
@@ -1,30 +1,146 @@
+  get mainnet() {
+    const _providers = getProvider(SupportedChainId.MAINNET)
+
+    if (_providers.length > 0) {
```
isn´t this strange to control a casae where we didn´t add the provider for mainnet. Since we do this in a static function, why do we need to do it?

anxolin replied:
you actually did sth similar for the URLs in coinbase, and we asusme there is at list one URL (i feel is a similar case, but you did sth different)

elena-zh commented:
Should we still keep it opened?
